README.md
Metroextractor Chef
===================
![Build Status](https://circleci.com/gh/mapzen/chef-metroextractor.svg?style=svg)

Attribution
-----------
Heavily influenced by Mike Migurski's [Extractotron](https://github.com/migurski/Extractotron/)

What does it do?
----------------
Downloads the latest planet in pbf format.
Produces, via osmosis, a number of metro extracts in both pbf and bz2 formats.
Produces shape files in both imposm and osm2pgsql formats.

Projections
-----------
GeoJSON is generated via ogr2ogr and output in CRS:84 (e.g. EPSG:4326) projection.

Imposm shapfiles are generated using default projections. We use imposm3, which currently means:
  * EPSG:3857

What hardware do I need?
------------------------
On AWS, a c3.8xl (60GB RAM and at least 75GB of fast disk), will result in the entire process taking ~4 hours.

If you'd rather not go that route, you're in luck. We'll be producing extracts weekly, and you can find
them here: [Mapzen Metro Extracts](https://mapzen.com/metro-extracts/)

Contributing
------------
Extract creation is driven off the cities.json file found here: [metroextractor-cities](https://github.com/mapzen/metroextractor-cities).
If you'd like to add an extract, you can either update the file directly if you feel comfortable doing so and submit a pull request,
or open a GitHub issue and we'll look into doing it for you.

If you'd like to update any other code, please fork, update, write specs for your changes and submit a pull.

Usage
-----
    include_recipe 'metroextractor'

Supported Platforms
-------------------
Tested and supported on the following platforms:

* Ubuntu 12.04LTS
* Ubuntu 14.04LTS

Requirements
------------
* Chef >= 11.4

Attributes
----------
* see [attributes/default.rb](https://github.com/mapzen/chef-metroextractor/blob/master/attributes/default.rb)

Dependencies
-----------
apt, ark, osm2pgsql, osmosis postgresql, python, user

Vagrant Environment
===================

Installation
------------
    vagrant plugin install vagrant-berkshelf 
    bundle install
    berks install
    vagrant up
    vagrant ssh

#### Note
* running a planet download and then processing extracts on your local machine might be hit or miss (likely miss)

#### I don't like Vagrant
* well then sir, provision an Ubuntu12.04 LTS system with the provider of your choice, and then bootstrap with chef-solo:
    `knife solo bootstrap root@${host} -r 'recipe[metroextractor]'`
* and re-cook with the following:
    `knife solo cook root@${host} -r 'recipe[metroextractor]'`
* alternatively, you can add the metroextractor cookbook to your chef server and wrap it as you see fit

License and Authors
-------------------
License: GPL
Authors: grant@mapzen.com


CHANGELOG.md
metroextractor changelog
========================

0.14.0
------
* replace osmosis with vex

0.13.8
------
* replace remote_file for planet download with wget: remote_file is too slow and flaky for files this large

0.13.7
------
* lower default heap size to 2G per osmosis processes

0.13.6
------
* make the planet download trigger off the md5 remote_file resource. If the md5 hasn't changed, we won't
  go through the trouble of downloading the planet again.
* calculate the md5 of the downloaded planet file and compare it against the provided md5sum

0.13.5
------
* :create the planet_remote file to allow use on stateful systems (i.e. non instance store)
* don't back up the planet file when we get a new one

0.13.4
------
* replace chef-rewind with resource override

0.13.3
------
* imposm parallel job tuning

0.13.2
------
* PG tuning

0.13.1
------
* break out parallel_jobs into two attributes: imposm_jobs and osm2pgsql_jobs, for more granular tuning

0.13.0
------
* update shapes and extracts to use gnu parallel: provide better logging
  and better error handling.

0.12.0
------
* go back to using remote_file to fetch the planet pbf

0.11.11
-------
* include branch in metroextractor-cities git pull, to allow override

0.11.10
-------
* don't re-sync the cities repo if we've already cloned it once, to avoid pulling in changes we
  aren't potentially expecting if we abort and re-start a run

0.11.9
------
* slice count

0.11.8
------
* drop the pg db during a re-run

0.11.7
------
* fewer slices

0.11.6
------
* pg max connections

0.11.5
------
* increase checkpoint_segments

0.11.4
------
* drop osm2pgsql num procs

0.11.3
------
* set slice count to cpu total

0.11.2
------
* read pbf rather than bz2 for osm2pgsql
* back to smaller slice size
* increase workers for osm2pgsql
* increase node cache for osm2pgsql

0.11.1
------
* use larger slices

0.11.0
------
* process shapes (osm2psql) in batches (default 8)

0.10.6
------
* node cache for osmosis

0.10.5
------
* fix osmosis heap setup

0.10.4
------
* allow download to linger for up to 2 hours

0.10.3
------
* limit osmosis heap size (per instance) via .osmosis conf file

0.10.2
------
* download url change

0.10.1
------
* osmosis buffer size

0.10.0
------
* big bump for osmosis changes

0.9.5
-----
* bug fixes

0.9.4
-----
* experiment with multiple osmosis processes, one per region

0.9.3
-----
* more workers for osmosis

0.9.2
-----
* remove hardcoded db/user/password in osm2pgsql.sh

0.9.1
-----
* need to support both imposm2/3 in osm2pgsql.sh

0.9.0
-----
* use imposm3
* TODO: specs for all the variations in setup.rb

0.8.4
-----
* test osm2pgsql.sh with osm2pgsql/imposm commands being run in series rather than parallel (hangs)

0.8.3
-----
* clean up imposm install logic

0.8.1
-----
* clean up pkg installs for 12.04 vs 14.04
* specs

0.8.0
-----
* tested on Ubuntu 14.04LTS
* on 12.04, continue installing imposm via pip, use pkg on 14.04

0.7.13
------
* don't merge osm2pgsql geojson this way. Disable for now.

0.7.12
------
* remove unnecessary 'rm' from cleanup in osm2pgsql.sh

0.7.11
------
* use DROP ... CASCADE

0.7.10
-----
* use execute rather than bash for planet download
* use --quiet

0.7.9
-----
* download planet with wget, remote_file is insanely slow with large files

0.7.8
-----
* speed up osmosis: set workers, buffers

0.7.7
-----
* remove compress=none from osmosis, as it doesn't appear to have any real effect
  on performance (speed), so might as well keep file size down

0.7.6
-----
* be explicit about modes
* set correct mode for merge script

0.7.5
-----
* override the source projection (-s_srs epsg:900913) for imposm geojson creation
  * should resolve issue https://github.com/mapzen/metroextractor-cities/issues/17

0.7.4
-----
* set compress=none in osmosis output. Per the docs, files should write twice as fast
  at the expense of 2x space requirement

0.7.3
-----
* add omitmetadata to osmosis cmd line

0.7.2
-----
* DROP VIEW should be DROP TABLE in osm2pgsql.sh

0.7.1
-----
* merge osm2pgsql geojson output into a single file

0.7.0
-----
* install imposm via pip to get latest 2.x release, in hopes of 
  resolving an ongoing imposm bug (hangs)

0.6.9
-----
* clean up osm2pgsql.sh

0.6.8
-----
* more fixes for geojson output
* add geojson for imposm shp files

0.6.7
-----
* bug fix for geojson output

0.6.6
-----
* bug fix in shapes.rb for lockfile

0.6.5
-----
* lock the planet download

0.6.4
-----
* osm2pgsql.sh fix: function, not funtion (!)

0.6.3
-----
* add geojson output

0.6.2
-----
* add clipIncompleteEntities=true to bounding box statements in osmosis.sh 
  to avoid dangling references

0.6.1
-----
* use localhost in osm2pgsql.sh

0.6.0
-----
* break out cities.json into its own repository

0.5.1
-----
* brisbane, perth, darwin...

0.5.0
-----
* fixes to allow easier overrides

0.4.2
-----
* bug fix in postgres template

0.4.1
-----
* restart postgres immediately

0.4.0
-----
* rewind the default postgres template
* apply more postgres tuning

0.3.2
-----
* turn off autovacuum
* checkpoint segments

0.3.1
-----
* add avignon to cities.json

0.3.0
-----
* override default postgres data dir to avoid running out of space

0.2.1
-----
* use array in osm2pgsql.sh

0.2.0
-----
* drive osm2pgsql/osmosis off a single json config file: cities.txt

0.1.1
-----
* stable


recipes/setup.rb
#
# Cookbook Name:: metroextractor
# Recipe:: setup
#

%w(
  osm2pgsql::default
).each do |r|
  include_recipe r
end

# packages for 12.04 and 14.04
#
%w(
  git
  build-essential
  gdal-bin
  parallel
  zip
).each do |p|
  package p
end

# packages for compiling/installing imposm on 12.04.
#   use pip on <= 12.04, pkg when > 12.04
#
%w(
  libtokyocabinet-dev
  libprotobuf-dev
  protobuf-c-compiler
  protobuf-compiler
  python-dev
  python-pip
  zlib1g-dev
).each do |p|
  package p do
    action :install
    only_if { platform?('ubuntu') && node[:platform_version] == '12.04' && node[:metroextractor][:imposm][:major_version] == 'imposm2' }
  end
end

python_pip 'imposm' do
  version '2.5.0'
  only_if { platform?('ubuntu') && node[:platform_version] <= '12.04' && node[:metroextractor][:imposm][:major_version] == 'imposm2' }
end

package 'imposm' do
  action :install
  only_if { platform?('ubuntu') && node[:platform_version] > '12.04' && node[:metroextractor][:imposm][:major_version] == 'imposm2' }
end

ark 'imposm3' do
  owner         'root'
  url           node[:metroextractor][:imposm][:url]
  version       node[:metroextractor][:imposm][:version]
  prefix_root   node[:metroextractor][:imposm][:installdir]
  has_binaries  ['imposm3']
  only_if       { node[:metroextractor][:imposm][:major_version] == 'imposm3' }
end

# vex
#
ark 'vex' do
  owner        'root'
  url          node[:metroextractor][:vex][:url]
  version      node[:metroextractor][:vex][:version]
  prefix_root  node[:metroextractor][:vex][:installdir]
  has_binaries ['vex']
  notifies     :run, 'execute[build vex]', :immediately
end

execute 'build vex' do
  action  :nothing
  cwd     "#{node[:metroextractor][:vex][:installdir]}/vex-#{node[:metroextractor][:vex][:version]}"
  command 'make'
end

# scripts basedir
#
directory node[:metroextractor][:setup][:scriptsdir] do
  owner node[:metroextractor][:user][:id]
end

# cities
#
git "#{node[:metroextractor][:setup][:scriptsdir]}/metroextractor-cities" do
  action      :sync
  repository  node[:metroextractor][:setup][:cities_repo]
  revision    node[:metroextractor][:setup][:cities_branch]
  user        node[:metroextractor][:user][:id]
end

link "#{node[:metroextractor][:setup][:scriptsdir]}/cities.json" do
  to "#{node[:metroextractor][:setup][:scriptsdir]}/metroextractor-cities/cities.json"
end

# scripts
#
template "#{node[:metroextractor][:setup][:scriptsdir]}/extracts.sh" do
  owner   node[:metroextractor][:user][:id]
  source  'extracts.sh.erb'
  mode    0755
end

template "#{node[:metroextractor][:setup][:scriptsdir]}/shapes.sh" do
  owner   node[:metroextractor][:user][:id]
  source  'shapes.sh.erb'
  mode    0755
end

cookbook_file "#{node[:metroextractor][:setup][:scriptsdir]}/osm2pgsql.style" do
  owner   node[:metroextractor][:user][:id]
  source  'osm2pgsql.style'
  mode    0644
end

cookbook_file "#{node[:metroextractor][:setup][:scriptsdir]}/merge-geojson.py" do
  owner   node[:metroextractor][:user][:id]
  source  'merge-geojson.py'
  mode    0755
end

# directories
#
directory node[:metroextractor][:vex][:db] do
  recursive true
  owner     node[:metroextractor][:user][:id]
  not_if    { node[:metroextractor][:vex][:db] == 'memory' }
end

%w(ex shp logs).each do |d|
  directory "#{node[:metroextractor][:setup][:basedir]}/#{d}" do
    recursive true
    owner     node[:metroextractor][:user][:id]
  end
end
